lib/ain-ocean/src/api/poolpairs.rs

```diff
@@ -426,36 +416,152 @@ async fn list_pool_swaps_verbose(
 //     )
 // }
 
-// #[ocean_endpoint]
-// async fn get_swappable_tokens(Path(SwappableTokens { token_id }): Path<SwappableTokens>) -> String {
-//     format!("Swappable tokens for token id {}", token_id)
-// }
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct AllSwappableTokensResponse {
+    from_token: TokenIdentifier,
+    swappable_tokens: Vec<TokenIdentifier>,
+}
 
-// #[ocean_endpoint]
-// async fn get_best_path(
-//     Query(BestPath {
-//         from_token_id,
-//         to_token_id,
-//     }): Query<BestPath>,
-// ) -> String {
-//     format!(
-//         "Best path from token id {} to {}",
-//         from_token_id, to_token_id
-//     )
-// }
+#[ocean_endpoint]
+async fn get_swappable_tokens(
+    Path(token_id): Path<String>,
+    Extension(ctx): Extension<Arc<AppContext>>,
+) -> Result<Response<AllSwappableTokensResponse>> {
+    sync_token_graph_if_empty(&ctx).await?;
+
+    let mut token_ids: HashSet<u32> = HashSet::new();
+
+    fn recur(ctx: &Arc<AppContext>, mut token_ids: HashSet<u32>, token_id: u32) -> HashSet<u32> {
+        if token_ids.contains(&token_id) {
+            return token_ids
+        };
+        token_ids.insert(token_id);
+        let graph = ctx.services.token_graph.lock().clone();
+        let edges = graph.edges(token_id).collect::<Vec<_>>();
+        for edge in edges {
+            token_ids = recur(ctx, token_ids.clone(), edge.0);
+            token_ids = recur(ctx, token_ids.clone(), edge.1);
+        }
+        token_ids
+    }
```

Comment:
Do we have to both take a mut ref and return the HashSet ?
We can lock only once.
We don't need to clone the set.

Something like :

```
let mut token_ids = HashSet::new();

    fn recur(ctx: &Arc<AppContext>, graph: &UnGraphMap<u32, String>, token_ids: &mut HashSet<u32>, token_id: u32) {
        if token_ids.contains(&token_id) {
            return
        };
        token_ids.insert(token_id);
        let edges = graph.edges(token_id).collect::<Vec<_>>();
        for edge in edges {
            recur(ctx, graph, token_ids, edge.0);
            recur(ctx, graph, token_ids, edge.1);
        }
    }

    {
        let graph = ctx.services.token_graph.lock().clone();
        recur(&ctx, &graph, &mut token_ids, token_id.parse::<u32>()?);
    }

    token_ids.remove(&token_id.parse::<u32>()?);